File: legacy/isma-core/src/main/java/ru/nstu/isma/core/common/SimulationResult.java
package ru.nstu.isma.core.common;

import java.util.*;
import java.util.stream.Collectors;

/**
 * Created by Bessonov Alex
 * on 12.10.2014.
 */
public class SimulationResult {
    protected ArrayList<Double> x = new ArrayList<>();
    protected Map<String, ArrayList<Double>> y = new HashMap<>();

    public void addEquation(String name) {
        if (y.containsKey(name))
            return;
        y.put(name, new ArrayList<>());
    }

    public void addX(Double x) {
        this.x.add(x);
    }

    public void addY(String name, Double y) {
        assertEquationContains(name);
        this.y.get(name).add(y);
    }

    public List<Point> getPoints(String name) {
        assertEquationContains(name);
        assertResultCorrect(name);
        List<Point> res = new LinkedList<>();
        for (int i = 0; i < x.size(); i++)
            res.add(new Point(x.get(i), y.get(name).get(i)));
        return res;
    }

    public List<String> equations() {
        return y.keySet().stream().collect(Collectors.toList());
    }

    private void assertEquationContains(String name) {
        if (!this.y.containsKey(name))
            throw new RuntimeException("SimulationResult don't contain equation " + name);
    }

    private void assertResultCorrect(String name) {
        if (y.get(name).size() != x.size())
            throw new RuntimeException("SimulationResult have different size for equation " + name);
    }
}

File: isma-next-core/src/main/kotlin/ru/nstu/isma/next/core/common/SimulationResult.kt
package ru.nstu.isma.next.core.common

import java.util.*
import java.util.stream.Collectors

/**
 * Created by Bessonov Alex
 * on 12.10.2014.
 */
class SimulationResult {
    protected var x = ArrayList<Double>()
    protected var y: MutableMap<String, ArrayList<Double>> = HashMap()
    fun addEquation(name: String) {
        if (y.containsKey(name)) return
        y[name] = ArrayList()
    }

    fun addX(x: Double) {
        this.x.add(x)
    }

    fun addY(name: String, y: Double) {
        assertEquationContains(name)
        this.y[name]!!.add(y)
    }

    fun getPoints(name: String): List<Point> {
        assertEquationContains(name)
        assertResultCorrect(name)
        val res: MutableList<Point> = LinkedList()
        for (i in x.indices) res.add(Point(x[i], y[name]!![i]))
        return res
    }

    fun equations(): List<String> {
        return y.keys.stream().collect(Collectors.toList())
    }

    private fun assertEquationContains(name: String) {
        if (!y.containsKey(name)) throw RuntimeException("SimulationResult don't contain equation $name")
    }

    private fun assertResultCorrect(name: String) {
        if (y[name]!!.size != x.size) throw RuntimeException("SimulationResult have different size for equation $name")
    }
}